libs/ballot-interpreter-nh/src/layout.ts

```diff
@@ -0,0 +1,298 @@
+    frontGridPositions,
+    ({ contestId }) => contestId
+  );
+  const backGridPositionsByContest = groupBy(
+    backGridPositions,
+    ({ contestId }) => contestId
+  );
+
+  function clampRow(row: number): number {
+    return Math.max(0, Math.min(row, geometry.gridSize.height - 1));
+  }
+
+  function clampColumn(column: number): number {
+    return Math.max(0, Math.min(column, geometry.gridSize.width - 1));
+  }
+
+  for (const [i, gridPositionsByContestId] of [
+    frontGridPositionsByContest,
+    backGridPositionsByContest,
+  ].entries()) {
+    const contests: BallotPageContestLayout[] = [];
+
+    for (const gridPositions of gridPositionsByContestId.values()) {
+      const options: BallotPageContestOptionLayout[] = [];
+
+      for (const gridPosition of Array.from(gridPositions)) {
+        const optionTopLeft = timingMarkGrid.rows[
+          clampRow(gridPosition.row + optionGridOriginOffset.y)
+        ]?.[
+          clampColumn(gridPosition.column + optionGridOriginOffset.x)
+        ] as Point;
+        const optionBottomRight = timingMarkGrid.rows[
+          clampRow(
+            gridPosition.row + optionGridOriginOffset.y + optionGridSize.height
+          )
+        ]?.[
+          clampColumn(
+            gridPosition.column +
+              optionGridOriginOffset.x +
+              optionGridSize.width
+          )
+        ] as Point;
+        const ovalCenter = timingMarkGrid.rows[gridPosition.row]?.[
+          gridPosition.column
+        ] as Point;
+
+        const optionBounds = makeRect({
+          minX: optionTopLeft.x,
+          minY: optionTopLeft.y,
+          maxX: optionBottomRight.x,
+          maxY: optionBottomRight.y,
+        });
+        const targetBounds = makeRect({
+          minX: Math.round(ovalCenter.x - geometry.ovalSize.width / 2),
+          minY: Math.round(ovalCenter.y - geometry.ovalSize.height / 2),
+          maxX: Math.round(ovalCenter.x + geometry.ovalSize.width / 2),
+          maxY: Math.round(ovalCenter.y + geometry.ovalSize.height / 2),
+        });
+        options.push({
+          bounds: optionBounds,
+          target: {
+            bounds: targetBounds,
+            inner: targetBounds,
+          },
+        });
+      }
+
+      const minX = Math.min(...options.map((o) => o.bounds.x));
+      const minY = Math.min(...options.map((o) => o.bounds.y));
+      const maxX = Math.max(
+        ...options.map((o) => o.bounds.x + o.bounds.width - 1)
+      );
+      const maxY = Math.max(
+        ...options.map((o) => o.bounds.y + o.bounds.height - 1)
+      );
+      const bounds = makeRect({ minX, minY, maxX, maxY });
+
+      contests.push({
+        options,
+        bounds,
+        corners: [
+          { x: bounds.x, y: bounds.y },
+          { x: bounds.x + bounds.width - 1, y: bounds.y },
+          { x: bounds.x, y: bounds.y + bounds.height - 1 },
+          { x: bounds.x + bounds.width - 1, y: bounds.y + bounds.height - 1 },
+        ],
```

Comment:
If `corners` is just a function of `bounds`, why specify both?

Reply:
The `BallotPageLayout` type applies not only to templates, where this is true, but also to scanned images where it may not be.